cmd/podman/logs.go

```diff
@@ -95,52 +105,116 @@ func logsCmd(c *cli.Context) error {
 		return err
 	}
 
-	logs := make(chan string)
-	go func() {
-		err = getLogs(ctr, logs, opts)
-	}()
-	printLogs(logs)
+	file, err := os.Open(ctr.LogPath())
+	if err != nil {
+		return errors.Wrapf(err, "unable to read container log file")
+	}
+	defer file.Close()
+	reader := bufio.NewReader(file)
+	if opts.follow {
+		followLog(reader, opts)
+	} else {
+		dumpLog(reader, opts)
+	}
 	return err
 }
 
-// getLogs returns the logs of a container from the log file
-// log file is created when the container is started/ran
-func getLogs(container *libpod.Container, logChan chan string, opts logOptions) error {
-	defer close(logChan)
-
-	seekInfo := &tail.SeekInfo{Offset: 0, Whence: 0}
+func followLog(reader *bufio.Reader, opts logOptions) error {
+	var cacheOutput []string
+	firstPass := false
 	if opts.tail > 0 {
-		// seek to correct position in log files
-		seekInfo.Offset = int64(opts.tail)
-		seekInfo.Whence = 2
+		firstPass = true
 	}
-
-	t, err := tail.TailFile(container.LogPath(), tail.Config{Follow: opts.follow, ReOpen: false, Location: seekInfo})
-	for line := range t.Lines {
-		if since, err := logSinceTime(opts.sinceTime, line.Text); err != nil || !since {
+	// We need to read the entire file in here until we reach EOF
+	// and then dump it out in the case that the user also wants
+	// tail output
+	for {
+		line, err := reader.ReadString('\n')
+		if err == io.EOF && opts.follow {
+			if firstPass {
+				firstPass = false
+				cacheLen := int64(len(cacheOutput))
+				start := int64(0)
+				if cacheLen > int64(opts.tail) {
+					start = cacheLen - int64(opts.tail)
+				}
+				for i := start; i < cacheLen; i++ {
+					printLine(cacheOutput[i], opts)
+				}
+				continue
+			}
+			time.Sleep(1 * time.Second)
 			continue
 		}
-		logMessage := line.Text[secondSpaceIndex(line.Text):]
-		logChan <- logMessage
+		// exits
+		if err != nil {
+			break
+		}
+		if firstPass {
+			cacheOutput = append(cacheOutput, line)
+			continue
+		}
+		printLine(line, opts)
 	}
-	return err
+	return nil
 }
 
-func printLogs(logs chan string) {
-	for line := range logs {
-		fmt.Println(line)
+func dumpLog(reader *bufio.Reader, opts logOptions) error {
+	output := readLog(reader, opts)
+	for _, line := range output {
+		printLine(line, opts)
 	}
+
+	return nil
+}
+
+func readLog(reader *bufio.Reader, opts logOptions) []string {
+	var output []string
+	for {
+		line, err := reader.ReadString('\n')
+		if err != nil {
+			break
```

Comment:
Should we check if this is actually EOF or something else blew up?

Reply:
same as above, imo it is fine as is